### ISSUE TITLE
[infra] Precompile libFuzzer

### ISSUE THREAD

**Issue by jonathanmetzman** [google/oss-fuzz#4677]
Currently, every time one compiles libfuzzer targets for a project, libFuzzer is rebuilt using the sanitizer one is using.
This step takes about 23 seconds on my machine.
I find it annoying as it increases the size of the feedback loop when I'm debugging stuff.
It also slows down CIFuzz, OSS-Fuzz's CI, our integration tests that build fuzzers and pretty much everything else that builds fuzzers using OSS-Fuzz.

Instead we could precompile libFuzzer in base-builder with sanitizers, and simply copy the desired build into /usr/lib/libFuzzingEngine.a.

Actually, it's worth examining if we need to compile libFuzzer with sanitizers at all when using UBSAN and ASAN. 
Unlike MSAN (and dfsan I think), ASAN instrumented binaries do not need to be 100% instrumented with ASAN.
We already have two (i386 and x64) uninstrumented builds of libFuzzer that we build with clang. We can maybe just copy these to /usr/lib/libFuzzingEngine.a in libFuzzer + ASAN/UBSAN builds. Not only will this will speed up compilation for those builds (which are the most common), it will also speed up fuzzing (since instrumenting libFuzzer slows it down) without slowing down building of base-builder or increasing the image size at all.

Alternatively, we could also speed up libFuzzer compilation by using different processes for compiling each .cpp file, just like libFuzzer's build.sh script does.

**Comment by inferno-chromium** [comment]
++++++++++++++++++++++++++++++++++++++++++++++++++++1, this is so annoying.

**Comment by jonathanmetzman** [comment]
Actually, the clang-bundled libFuzzer build should in theory be usable with MSAN targets.

**Comment by jonathanmetzman** [comment]
This is probably blocked by https://github.com/google/oss-fuzz/issues/2375

**Comment by jonathanmetzman** [comment]
Fixed by #4682.

**Comment by jonathanmetzman** [comment]
There are a few projects that broke for the first time after #4682 was landed.
Most seem unrelated:
[libgd](https://oss-fuzz-build-logs.storage.googleapis.com/log-d4c6119d-5f79-4808-9466-6f67b11c3b51.txt)
[libmediaart](https://oss-fuzz-build-logs.storage.googleapis.com/log-7a55023a-1587-4255-baa3-88d9b2568a97.txt)
[monero](https://oss-fuzz-build-logs.storage.googleapis.com/log-f6de7849-8ad6-4de1-a476-8c4a2ff6d2aa.txt)
[proxygen](https://oss-fuzz-build-logs.storage.googleapis.com/log-597b2190-466b-4c59-a0dd-03f4f932bdd5.txt)
[cppcheck](https://oss-fuzz-build-logs.storage.googleapis.com/log-0b42cb64-e42c-47b9-b44d-07b113313f78.txt)
Only one seems to be caused by this patch:
[casync](https://oss-fuzz-build-logs.storage.googleapis.com/log-7850ae9a-f0c0-431c-addd-cad8bc705944.txt)
The line that fails is
```
Step #4: clang++  -o fuzz-compress fuzz-compress.p/test_fuzz_fuzz-compress.c.o -L/usr/lib/clang/12.0.0/lib/linux -Wl,--as-needed -Wl,--allow-shlib-undefined -Wl,--start-group src/libshared.a -lacl /usr/lib/x86_64-linux-gnu/liblzma.so /usr/lib/x86_64-linux-gnu/libselinux.so /usr/lib/x86_64-linux-gnu/libz.so -lm /usr/lib/x86_64-linux-gnu/libssl.so /usr/lib/x86_64-linux-gnu/libcrypto.so -pthread -lFuzzingEngine -Wl,--end-group
```
It fails with following errors (condensed):
```
Step #4: fuzz-compress.c:(.text+0x2): undefined reference to `__asan_init'
Step #4: fuzz-compress.c:(.text+0x7): undefined reference to `__asan_version_mismatch_check_v8'
Step #4: fuzz-compress.c:(.text+0x18): undefined reference to `__asan_register_globals'
Step #4: fuzz-compress.p/test_fuzz_fuzz-compress.c.o: In function `asan.module_dtor':
Step #4: fuzz-compress.c:(.text+0x2e): undefined reference to `__asan_unregister_globals'
Step #4: fuzz-compress.p/test_fuzz_fuzz-compress.c.o: In function `LLVMFuzzerTestOneInput':
Step #4: /work/build/../../src/casync/test/fuzz/fuzz-compress.c:18: undefined reference to `__asan_option_detect_stack_use_after_return'
Step #4: /work/build/../../src/casync/test/fuzz/fuzz-compress.c:18: undefined reference to `__asan_stack_malloc_3'
```
Basically i think the issue is that they aren't using CXXFLAGS to link the fuzz target. Previously this worked because builtin libFuzzer had ASAN linked in. The fix is probably to link with CXXFLAGS as well.
